bpo-47049: Fix incorrect shutil.copytree() behaviour with symlinks

vajdaz opened the pull request:
shutil.copytree incorrectly does not copy symlink contents if called
with symlink=False and ignore_dangling_symlinks=True.

The wrong behaviour can be reproduced like this:
```
$ tree
.
└── a
    ├── a.txt
    └── b
        └── a.txt -> ../a.txt
```
```$ python3 -c "import shutil;shutil.copytree('a/b', 'c', symlinks=False, ignore_dangling_symlinks=True)"```

As a result directoy c will be created but it will remain empty.
Expected result is a file c/a.txt with the contents of a/b/a.txt.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-47049](https://bugs.python.org/issue47049) -->
https://bugs.python.org/issue47049
<!-- /issue-number -->

vajdaz committed:
Fix incorrect shutil.copytree() with symlinks

shutil.copytree incorrectly does not copy symlink contents if called
with symlink=False and ignore_dangling_symlinks=True.

The wrong behaviour can be reproduced like this:

$ tree
.
└── a
    ├── a.txt
    └── b
        └── a.txt -> ../a.txt

$ python3 -c "import shutil;shutil.copytree('a/b', 'c', symlinks=False, ignore_dangling_symlinks=True)"

As a result directoy c will be created but it will remain empty.
Expected result is a file c/a.txt with the contents of a/b/a.txt.

arhadthedev commented on Lib/shutil.py:
```diff
@@ -483,6 +483,12 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
                 else:
+                    # If the symlink is relative, linkto has to be normalized,
+                    # otherwise os.path.exists() will incorrectly report that
+                    # the link is dangling when the current directoy is not
```
```suggestion
                    # the link is dangling when the current directory is not
```

vajdaz committed:
Correct typo in the comment

Co-authored-by: Oleg Iarygin <dralife@yandex.ru>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

vajdaz committed:
Update 2022-03-17-20-58-28.bpo-47049.DgTJ0k.rst

Fix typo

vajdaz committed:
Add unit test to the fix of issue 47049

The test creates following filesystem structure:

src_dir
├── test.txt
└── subdir
    ├── doesnotexist.txt -> IDONOTEXIST
    └── test.txt -> ../test.txt

Then it uses copytree with symlinks=False and ignore_dangling_symlinks=True
As a result, in the destination file structure the symlink should become
a file and its contents should be the same as that of the original source
file that the symlink was pointing to. The symlink doesnotexist.txt
should be ignored, no error should occur, and no copy should appear in
the destination file tree.

The test revaled another issue with the _copytree implementation, which
was fixed, too.

vajdaz committed:
Fix previous commit

The unit test in the prevous commit did not test the case that failed
prior to the fix, it also succeeded with Python versions containing
the bug. Now it tests exaclty the same situation that is described in
issue 47049.

MaxwellDupre reviewed (approved):
LGTM: module and given tests run ok.
However, there is a lot of comments Lib/test/test_shutil.py esp lines 771-778. Most comment for tests are only a few lines.
You already hae ref'ed the Issue 47049 so people can go look that up if they want or check the PR. 
Hence, you description lines 779 onwards could be a little shorter too.

vajdaz committed:
Cut test comments

As a result of the code review the test comments were reduced.

vajdaz committed:
Merge branch 'main' into fix-copytree-symlink-plus-ignore-dangling

MaxwellDupre reviewed (approved):
Better but still a few extra words.

vajdaz committed:
Cut more test comments
